floris/tools/optimization/layout_optimization/layout_optimization_pyoptsparse.py

```diff
@@ -0,0 +1,200 @@
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations under
+# the License.
+
+# See https://floris.readthedocs.io for documentation
+
+
+import numpy as np
+import matplotlib.pyplot as plt
+from shapely.geometry import Point
+from scipy.spatial.distance import cdist
+
+from .layout_optimization_base import LayoutOptimization
+
+class LayoutOptimizationPyOptSparse(LayoutOptimization):
+    def __init__(
+        self,
+        fi,
+        boundaries,
+        min_dist=None,
+        freq=None,
+        solver=None,
+        optOptions=None,
+    ):
+        super().__init__(fi, boundaries, min_dist=min_dist, freq=freq)
+        self._reinitialize(solver=solver, optOptions=optOptions)
+
+    def _reinitialize(self, solver=None, optOptions=None):
+        try:
+            import pyoptsparse
+        except ImportError:
+            err_msg = (
+                "It appears you do not have pyOptSparse installed. "
+                + "Please refer to https://pyoptsparse.readthedocs.io/ for "
+                + "guidance on how to properly install the module."
+            )
+            self.logger.error(err_msg, stack_info=True)
+            raise ImportError(err_msg)
+
+        # Insantiate ptOptSparse optimization object with name and objective function
+        self.optProb = pyoptsparse.Optimization('layout', self._obj_func)
+
+        self.optProb = self.add_var_group(self.optProb)
+        self.optProb = self.add_con_group(self.optProb)
+        self.optProb.addObj("obj")
+
+        if solver is not None:
+            self.solver = solver
+            print("Setting up optimization with user's choice of solver: ", self.solver)
+        else:
+            self.solver = "SLSQP"
+            print("Setting up optimization with default solver: SLSQP.")
+        if optOptions is not None:
+            self.optOptions = optOptions
+        else:
+            if self.solver == "SNOPT":
+                self.optOptions = {"Major optimality tolerance": 1e-7}
+            else:
+                self.optOptions = {}
+
+        exec("self.opt = pyoptsparse." + self.solver + "(options=self.optOptions)")
+
+    def _optimize(self):
+        if hasattr(self, "_sens"):
+            self.sol = self.opt(self.optProb, sens=self._sens)
+        else:
+            self.sol = self.opt(self.optProb, sens="CDR", storeHistory='hist.hist')
+        return self.sol
+
+    def _obj_func(self, varDict):
+        # Parse the variable dictionary
+        self.parse_opt_vars(varDict)
+
+        # Update turbine map with turbince locations
+        self.fi.reinitialize(layout=[self.x, self.y])
+        self.fi.calculate_wake()
+
+        # Compute the objective function
+        funcs = {}
+        funcs["obj"] = (
+            -1 * np.sum(self.fi.get_farm_power() * self.freq * 8760) / self.initial_AEP
```

Comment:
Could you use `fi.get_farm_AEP()` for this operation? This would simplify the calculation. You could even consider adding the `cut_in_wind_speed` and `cut_out_wind_speed` options that are in `fi.get_farm_AEP` to allow the user to constrain the wind speeds that are actually evaluated in the optimization.

Reply:
Updated to `fi.get_farm_AEP()`. I will need to think through passing in the cut in and cut out wind speeds more, as they would need defaults, and right now `floris_interface` doesn't handle a `None` for the cut in wind speed.